examples/seq2seq/seq2seq_trainer.py

```diff
@@ -0,0 +1,205 @@
+import logging
+import warnings
+from typing import Any, Dict, Optional, Tuple, Union
+
+import torch
+from packaging import version
+from torch import nn
+from torch.utils.data import DistributedSampler, RandomSampler
+
+from transformers import Trainer
+from transformers.file_utils import is_apex_available, is_torch_tpu_available
+from transformers.trainer import get_tpu_sampler
+
+
+_use_native_amp = False
+_use_apex = False
+
+# Check if Pytorch version >= 1.6 to switch between Native AMP and Apex
+if version.parse(torch.__version__) < version.parse("1.6"):
+    if is_apex_available():
+        from apex import amp
+    _use_apex = True
+else:
+    _use_native_amp = True
+    from torch.cuda.amp import autocast
+
+try:
+    from .utils import label_smoothed_nll_loss
+except ImportError:
+    from utils import label_smoothed_nll_loss
+
+
+logger = logging.getLogger(__name__)
+
+
+class Seq2SeqTrainer(Trainer):
+    def _get_train_sampler(self) -> Optional[torch.utils.data.sampler.Sampler]:
+        if isinstance(self.train_dataset, torch.utils.data.IterableDataset):
+            return None
+        elif is_torch_tpu_available():
+            return get_tpu_sampler(self.train_dataset)
+        else:
+            if self.args.sortish_sampler:
+                if self.args.n_gpu <= 1 and self.args.local_rank == -1:
+                    return self.train_dataset.make_sortish_sampler(self.args.per_device_train_batch_size)
+                else:
+                    warnings.warn("sortish_sampler is being ignored because n_gpu > 1")
+
+            return (
+                RandomSampler(self.train_dataset)
+                if self.args.local_rank == -1
+                else DistributedSampler(self.train_dataset)
+            )
+
+    # override to support label smoothing
+    def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]]) -> float:
```

Comment:
Seems very long for a simple use case. Might add the `compute_loss` to Trainer directly to more easily support this use case :-)

Reply:
Thanks Sylvain !
Adding it  to `Trainer` would be nice. Is `label_smoothing` also used for tasks other than seq2seq ?
LMK, I'll be happy to move to `Trainer` in this or separate PR.